presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -762,16 +758,13 @@ public PlanWithProperties visitJoin(JoinNode node, Context context)
             PlanWithProperties left;
             PlanWithProperties right;
 
-            boolean isCrossJoin = type == INNER && leftSymbols.isEmpty();
-            if ((distributedJoins && !isCrossJoin && !isScalar(node.getRight())) || (type == FULL) || (type == RIGHT)) {
-                // The implementation of full outer join only works if the data is hash partitioned. See LookupJoinOperators#buildSideOuterJoinUnvisitedPositions
-
+            if (node.getDistributionType().orElseThrow(() -> new IllegalStateException("distributionType not yet set")) == JoinNode.DistributionType.PARTITIONED) {
```

Comment:
Move the illegal state check outside of the `if` condition. Doing it inline makes the code harder to follow.